### PR TITLE
Enhancement/mf merge wizard into new ux

### PR DESCRIPTION
**Pull request by masonforest**
* I re-enabled some of the mocks so we can see the products marketplace with mocked data until we get the real data in there

### CODE REVIEW

**Comment by masonforest** [comment]
Note: The test suite isn't passing in the new UX yet so this branch won't pass either.

**Comment by chriskacerguis** [comment]
@masonforest can't auto-merge, can you take a look?

**Comment by AllenBW** [comment]
:+1:

File: src/client/app/components/tags/tag-field.directive.js [hunk 1]
```diff
@@ -259,7 +259,7 @@
 
       vm.newTag = '';
       vm.placeholder = '';
-      vm.tags = [];
+      vm.tags = vm.tags || [];
```
Comment (stackus) [inline src/client/app/components/tags/tag-field.directive.js]: Why was this added? The field shouldn't have any tags applied to it directly. They should be coming from the parent directive.

File: src/client/app/core/constants.js [hunk 2]
```diff
@@ -6,5 +6,7 @@
     .constant('lodash', _)
     .constant('jQuery', $)
     .constant('toastr', toastr)
+    .constant('WIZARD_AUTOSUBMIT', angular.element('meta[name="wizard-autosubmit"]').attr('content') == 'true')
```
Comment (stackus) [inline src/client/app/core/constants.js]: Don't mix configuration constants in with application core constants.

File: src/client/app/components/api/api.constants.js [hunk 3]
```diff
@@ -43,7 +43,7 @@
         'groupsById': '/groups/:id',
         'roles': '/roles',
         'rolesById': '/roles/:id',
-        'wizardQuestions': '/wizard_questions/:id'
+        'wizardQuestionsById': '/wizard_questions/:id'
```
Comment (stackus) [inline src/client/app/components/api/api.constants.js]: Consider dropping this file completely. It only caused developers to look into two files to get the complete picture of resources endpoints. Each value is only ever used in a single resource anyhow.

File: src/client/app/mock-http-backend/config.js [hunk 4]
```diff
@@ -5,6 +5,9 @@
     .run(mock);
 
   function mock($httpBackend) {
-    $httpBackend.whenGET(/^(?!\/api\/).+$/).passThrough();
+    $httpBackend.whenGET(/^(?!\/api\/mock\/).+$/).passThrough();
```
Comment (stackus) [inline src/client/app/mock-http-backend/config.js]: Feel like mocking things into special `/mock/` endpoints defeats the purpose of mocking things in the first place. Either the application intercepts the calls or it doesn't.

You're telling the application that all all `/api/mock` calls should go straight through to the real backend.

File: src/client/app/states/admin/wizard-questions/list/list.html [hunk 5]
```diff
@@ -2,16 +2,91 @@
   <div class="admin-products-list__body">
```
Comment (stackus) [inline src/client/app/states/admin/wizard-questions/list/list.html]: This is using styles and classes from the admin-products styles.

The UI does not match with the rest of the new UI.

File: src/client/app/states/admin/wizard-questions/list/list.html [hunk 6]
```diff
@@ -2,16 +2,91 @@
...
   </div>
-  <div class="admin-products-list__aside">
-    <div>
-      <product-type-show-modal
-        title="Add A Product"
-        message="Select the type of product you will be creating"
-        product-types="vm.productTypes"
-        on-ok="vm.createType(productType)"></product-type-show-modal>
+</content-header>
+<form class="form-horizontal padded horizontal-spanning">
```
Comment (stackus) [inline src/client/app/states/admin/wizard-questions/list/list.html]: Crutch classes `padded horizontal-spanning`.

File: src/client/app/states/admin/wizard-questions/list/list.html [hunk 7]
```diff
@@ -2,16 +2,91 @@
...
+  <div ng-repeat="question in vm.questions">
+    <hr>
+    <div class="row">
```
Comment (stackus) [inline src/client/app/states/admin/wizard-questions/list/list.html]: Form controls should be put into `form-control` divs, and the labels and input use appropriate form-* classes. These should be BEM'ed as well.

File: src/client/app/states/marketplace/marketplace.state.js [hunk 8]
```diff
@@ -14,7 +14,7 @@
   function getStates() {
     return {
       'marketplace': {
-        url: '/marketplace',
+        url: '/marketplace?tags',
```
Comment (stackus) [inline src/client/app/states/marketplace/marketplace.state.js]: Consider instead using ui-router params. See `app/states/admin/product/create/create.state.js` for an example.

File: src/client/app/states/marketplace/marketplace.state.js [hunk 9]
```diff
@@ -45,14 +45,18 @@
...
-  function StateController(logger, $q, VIEW_MODES, CatalogService, Tag, Compare, TAG_QUERY_LIMIT) {
+  function StateController(logger, $q, VIEW_MODES, CatalogService, Tag, Compare, TAG_QUERY_LIMIT, $stateParams) {
     var vm = this;
 
     var COMPARE_LIMIT = 4;
 
     vm.title = 'Marketplace';
 
-    vm.tags = [];
+    if($stateParams.tags) {
```
Comment (stackus) [inline src/client/app/states/marketplace/marketplace.state.js]: This should be moved into activate().

File: src/client/app/states/marketplace/marketplace.state.js [hunk 10]
```diff
@@ -45,14 +45,18 @@
...
+      vm.tags =  $stateParams.tags;
+    } else {
+      vm.tags = [];
```
Comment (stackus) [inline src/client/app/states/marketplace/marketplace.state.js]: More specifically this should be left here, and the existence check should be moved into activate.

File: src/client/app/states/wizard/multipage-wizard/multipage-wizard.state.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+  /** @ngInject */
+  function appRun(routerHelper, navigationHelper) {
+    routerHelper.configureStates(getStates());
+  }
+
+  function getStates() {
+    return {
+      'wizard.multipage': {
+        controller: multiPageWizardController,
```
Comment (stackus) [inline src/client/app/states/wizard/multipage-wizard/multipage-wizard.state.js]: Use `StateController`.

File: src/client/app/states/admin/wizard-questions/list/list.state.js [hunk 12]
```diff
@@ -16,9 +16,14 @@
       'admin.wizard-questions.list': {
         url: '', // No url, this state is the index of admin.products
         templateUrl: 'app/states/admin/wizard-questions/list/list.html',
-        controller: StateController,
+        controller: WizardQuestionsController,
```
Comment (stackus) [inline src/client/app/states/admin/wizard-questions/list/list.state.js]: Use `StateController`.

File: src/client/app/states/admin/wizard-questions/list/list.state.js [hunk 13]
```diff
@@ -32,16 +37,56 @@
...
 
-    vm.title = 'Admin Products List';
-    vm.activate = activate;
+    vm.questions = questions;
+    vm.question = new WizardQuestion({wizard_answers: [{}]});
+    vm.createQuestion = createQuestion;
+    vm.addAnswer = addAnswer;
 
-    activate();
+    _.each(questions, addAnswer);
```
Comment (stackus) [inline src/client/app/states/admin/wizard-questions/list/list.state.js]: `lodash.each()`

lodash will need to also be injected into the controller. 

JSCS and JSHint should be telling you that this is a violation.

File: src/client/app/states/wizard/multipage-wizard/multipage-wizard.state.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    var vm = this;
+
+    vm.nextQuestion = nextQuestion;
+    vm.projectId = $stateParams.projectId;
+    vm.question = question;
+    vm.resetWizard = resetWizard;
+    vm.tags = [];
+
+    function nextQuestion() {
+      vm.tags = _.union(vm.tags, vm.answer.tags_to_add);
```
Comment (stackus) [inline src/client/app/states/wizard/multipage-wizard/multipage-wizard.state.js]: `lodash.union()`

File: src/client/app/states/wizard/multiple_choice_directive.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
+(function() {
```
Comment (stackus) [inline src/client/app/states/wizard/multiple_choice_directive.js]: This is in the wrong place.

File: src/client/app/states/wizard/multiple_choice_directive.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  'use strict';
+
+  angular.module('app.components')
+    .directive('multipleChoice', MultipleChoiceDirective)
+
+  function MultipleChoiceDirective() {
+    return {
+      controller: MultipleChoiceDirectiveController,
+      link: handleModelUpdate,
```
Comment (stackus) [inline src/client/app/states/wizard/multiple_choice_directive.js]: Use `link`.

File: src/client/app/states/wizard/multiple_choice_directive.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+        action: "=",
+        model: "=",
+        options: "=",
+      },
+      templateUrl: 'app/states/wizard/multiple_choice.html',
+    }
+  };
+
+  function MultipleChoiceDirectiveController($scope, WIZARD_AUTOSUBMIT, WIZARD_MULTIPAGE) {
+    $scope.WIZARD_AUTOSUBMIT = WIZARD_AUTOSUBMIT;
```
Comment (stackus) [inline src/client/app/states/wizard/multiple_choice_directive.js]: There's no reason to put these values onto scope here. They're constants and are available always.

File: src/client/app/states/wizard/multiple_choice_directive.js [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+      scope: {
+        action: "=",
+        model: "=",
+        options: "=",
+      },
+      templateUrl: 'app/states/wizard/multiple_choice.html',
+    }
+  };
+
+  function MultipleChoiceDirectiveController($scope, WIZARD_AUTOSUBMIT, WIZARD_MULTIPAGE) {
```
Comment (stackus) [inline src/client/app/states/wizard/multiple_choice_directive.js]: Don't use `$scope`.

File: src/client/app/states/wizard/multiple_choice_directive.js [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    $scope.WIZARD_AUTOSUBMIT = WIZARD_AUTOSUBMIT;
+    $scope.WIZARD_MULTIPAGE = WIZARD_MULTIPAGE;
+  }
+
+  function handleModelUpdate(scope) {
+    if(scope.WIZARD_AUTOSUBMIT && scope.WIZARD_MULTIPAGE) {
+      scope.$watch('model', function(newValue){
```
Comment (stackus) [inline src/client/app/states/wizard/multiple_choice_directive.js]: This `$watch` doesn't appear to be necessary. It looks like you're trying to react to the value changing instead of taking reacting to user action.

File: src/client/app/states/wizard/multiple_choice_directive.js [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+  angular.module('app.components')
+    .directive('multipleChoice', MultipleChoiceDirective)
+
+  function MultipleChoiceDirective() {
+    return {
+      controller: MultipleChoiceDirectiveController,
+      link: handleModelUpdate,
+      restrict: 'E',
+      scope: {
+        action: "=",
```
Comment (stackus) [inline src/client/app/states/wizard/multiple_choice_directive.js]: Use `&?`. If you're going to be padding in function references use `&`. If the reference is optional add `?`.

File: src/client/app/states/wizard/singlepage-wizard/singlepage-wizard.state.js [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+  /** @ngInject */
+  function appRun(routerHelper, navigationHelper) {
+    routerHelper.configureStates(getStates());
+  }
+
+  function getStates() {
+    return {
+      'wizard.singlepage': {
+        controller: singlePageWizardController,
```
Comment (stackus) [inline src/client/app/states/wizard/singlepage-wizard/singlepage-wizard.state.js]: `StateController`

File: src/client/app/states/wizard/wizard.state.js [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+    .run(appRun);
+
+  /** @ngInject */
+  function appRun(routerHelper, navigationHelper, WIZARD_MULTIPAGE) {
+    routerHelper.configureStates(getStates(WIZARD_MULTIPAGE));
+  }
+
+  function getStates(WIZARD_MULTIPAGE) {
+    return {
+      'wizard': {
```
Comment (stackus) [inline src/client/app/states/wizard/wizard.state.js]: All of the child states of `wizard.` each resolve `questions`. That should instead be resolved here and not repeated for each child state.

File: src/client/app/states/wizard/wizard.state.js [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+        url: "/project/:projectId/wizard",
+        redirectTo: WIZARD_MULTIPAGE ? 'wizard.multipage' : 'wizard.singlepage',
+        template: '<ui-view></ui-view>'
```
Comment (stackus) [inline src/client/app/states/wizard/wizard.state.js]: This can be dropped.

File: src/client/app/components/api/api.constants.js (hunk 3 above)
Comment (masonforest) [inline src/client/app/components/api/api.constants.js]: I agree! :+1:  I was following the convention here.

Let's discuss this change as a team and make a separate PR to remove this file if that's what we decide to do.

File: src/client/app/components/tags/tag-field.directive.js (hunk 1 above)
Comment (stackus) [inline src/client/app/components/tags/tag-field.directive.js]: I slept on this and think I now know why this was added. The actual init should happen in `activate()` though.

I know I've mentioned activate a lot and should explain why. The reason to put init, or startup code into activate is because eventually when the code is upgraded to Angular 2.0 and it uses ES6, the activates should be 1:1 transforms to the class constructor.

Comment (masonforest) [inline src/client/app/components/tags/tag-field.directive.js]: Ah, gotchya. Thanks for the explanation: Fixed: [Move tags initialization to `activate()`](https://github.com/projectjellyfish/api/commit/3ac200183d11647d1340a3e7dc8a03b14a6bb19f)

File: src/client/app/mock-http-backend/config.js (hunk 4 above)
Comment (masonforest) [inline src/client/app/mock-http-backend/config.js]: Yeah, I partially mocked the API to I could test how tags would work with the Wizard (the tags API isn't done yet). You're right though having a partially mocked API is confusing and having this unmocked will push us to finish the API: [Revert partial API mocking](https://github.com/projectjellyfish/api/commit/3d3f5927e436ce0fcbe9f76fd5c9453558aed0a3)

File: src/client/app/states/admin/wizard-questions/list/list.html (hunk 6 above)
Comment (masonforest) [inline src/client/app/states/admin/wizard-questions/list/list.html]: The markup in the PR is just a place holder. I put minimal style on it to make it look not terrible. A designer will have to come through and make this pretty. At which point they can follow markup conventions, BEM, etc.

If we want I can try my hand a visual design but I wouldn't recommend it :flushed:

File: src/client/app/states/wizard/multipage-wizard/multipage-wizard.state.js (hunk 11 above)
Comment (masonforest) [inline src/client/app/states/wizard/multipage-wizard/multipage-wizard.state.js]: Oh nice. I didn't notice we were doing this. [Fixed here and elsewhere](https://github.com/projectjellyfish/api/commit/31bc528aa5a9c73a36a4564148881706cb5e2996).

File: src/client/app/states/wizard/wizard.state.js (hunk 22 above)
Comment (masonforest) [inline src/client/app/states/wizard/wizard.state.js]: Oh nice. I didn't know `resolves` worked across redirects. [Fixed](https://github.com/projectjellyfish/api/commit/3013fc30cc1b883ae556c4bbe44e685ad94cee5c)

File: src/client/app/states/wizard/wizard.state.js (hunk 23 above)
Comment (masonforest) [inline src/client/app/states/wizard/wizard.state.js]: I wouldn't think I would need this either but when I remove it the child template fails to load.

`template` is also set to `<ui-view></ui-view>` [here](https://github.com/projectjellyfish/api/blob/enhancement/JEL_%23601-Integrate-CM-UX-into-Jellyfis-UX/src/client/app/states/projects/projects.state.js#L19).

**Comment by masonforest** [comment]
https://www.dropbox.com/s/g6sa1kx6gy09cwh/jellyfish.mov?dl=0

File: src/client/app/states/wizard/multiple_choice_directive.js (hunk 17 above)
Comment (masonforest) [inline src/client/app/states/wizard/multiple_choice_directive.js]: Hmmmm. Is there a way to access constants in the views without putting them on `$scope`?

I need access to them [here](https://github.com/projectjellyfish/api/pull/670/files#diff-67572dd8c26424f37f7f491e0841d740R34).

Relevant stackoverflow answer: http://stackoverflow.com/a/19383833

File: src/client/app/components/multiple-choice/multiple-choice.html [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+        />
+      {{option.text}}
+    </label>
+    </div>
+  <input
+    class="btn btn-primary btn-lg pull-right"
+    type="submit"
+    value="Next"
+    ng-click="action()"
+    ng-hide="WIZARD_AUTOSUBMIT || !WIZARD_MULTIPAGE"
```
Comment (stackus) [inline src/client/app/components/multiple-choice/multiple-choice.html]: Instead of checking the constants you can do the check within the viewmodel and then do something like `vm.autoAdvance`

File: src/client/app/components/wizard/constants.js [hunk 25]
```diff
@@ -0,0 +1,7 @@
+(function() {
+  'use strict';
+
+  angular.module('app.core')
```
Comment (stackus) [inline src/client/app/components/wizard/constants.js]: module should be `app.components`.

File: src/client/app/components/multiple-choice/multiple-choice.html (hunk 24 above)
Comment (masonforest) [inline src/client/app/components/multiple-choice/multiple-choice.html]: Cool. I set `autoSubmit` and `multiPage` on `vm` and then checked those in the view. -> https://github.com/projectjellyfish/api/commit/06abe633b6f05b3bf8a4ffe0b59e6352673f0778
